bundles/tools.vitruv.change.propagation/src/tools/vitruv/change/propagation/impl/DefaultChangeRecordingModelRepository.java

```diff
@@ -0,0 +1,190 @@
+
+import org.eclipse.emf.common.util.URI;
+import org.eclipse.emf.ecore.EObject;
+import org.eclipse.emf.ecore.resource.Resource;
+
+/**
+ * A default implementation of a {@link PersistableChangeRecordingModelRepository}. It manages a {@link ResourceSet} for
+ * the models, which are automatically monitored for changes. It loads existing models calling
+ * {@link #loadExistingModels()} by loading the correspondence model at the URI specified in the constructor and
+ * resolving all models references in the correspondence model. If no such URI is specified, loading will have no
+ * effect.
+ */
+public class DefaultChangeRecordingModelRepository implements PersistableChangeRecordingModelRepository {
+	private static final Logger LOGGER = Logger.getLogger(DefaultChangeRecordingModelRepository.class);
+	private final ResourceSet modelsResourceSet;
+	private final InternalCorrespondenceModel correspondenceModel;
+	private final ChangeRecorder changeRecorder;
+	private final Path consistencyMetadataFolder;
+
+	private boolean isLoading = false;
+
+	/**
+	 * Creates a repository without a defined persistence for the correspondence model and using a temporary folder for
+	 * storing consistency metadata.
+	 * @throws IOException when no temporary metadata folder can be instantiated
+	 */
+	public DefaultChangeRecordingModelRepository() throws IOException {
+		this(null, Files.createTempDirectory(null));
+	}
+
+	/**
+	 * Creates a repository with the defined URI for the correspondence model and using the specified path for storing
+	 * metadata. The URI must be resolvable when performing load or save operations, or can be {@code null} is no
+	 * persistence of the correspondence model shall be managed.
+	 * @param correspondencesURI the URI of the correspondence model to use for saving and loading it, must be {@code null}
+	 * or valid
+	 * @param consistencyMetadataFolder the folder to store consistency metadata in, must not be {@code null}
+	 */
+	public DefaultChangeRecordingModelRepository(URI correspondencesURI, Path consistencyMetadataFolder) {
+		this.consistencyMetadataFolder = consistencyMetadataFolder;
+		this.modelsResourceSet = withGlobalFactories(new ResourceSetImpl());
+		this.correspondenceModel = createCorrespondenceModel(correspondencesURI);
+		this.modelsResourceSet.eAdapters().add(new ResourceRegistrationAdapter((resource) -> {
+			if (!isLoading) {
+				getCreateOrLoadModel(resource.getURI());
+			}
+		}));
+		this.changeRecorder = new ChangeRecorder(modelsResourceSet);
+	}
+
+	@Override
+	public void loadExistingModels() {
+		isLoading = true;
+		correspondenceModel.loadSerializedCorrespondences(modelsResourceSet);
+		isLoading = false;
+	}
+
+	@Override
+	public CorrespondenceModel getCorrespondenceModel() {
+		return correspondenceModel.getGenericView();
+	}
+
+	@Override
+	public URI getMetadataModelURI(String... metadataKey) {
+		Path metadataPath = consistencyMetadataFolder.resolve(String.join("_", metadataKey));
```

Comment:
Perhaps we should think about externalizing String constants like that, especially to avoid things like [this](https://github.com/vitruv-tools/Vitruv-Applications-ComponentBasedSystems/commit/000b41db69fc1dc0692a68027850fd1076df110a).

Reply:
I am not totally sure what you mean by "externalize". We may store it in a static final variable to have its meaning make explicit (that absolutely makes sense). But I have to admit that I do not know what the problem was in the referenced commit.

Reply:
The unicode character does not work for the windows CI build. Having all special characters grouped up in some sort of localization file could speed up the search but i do not think it is worth the effort. Static final variable sounds good.

Reply:
Hm, actually those characters should also work in Windows builds (at least they do on my system), but maybe there are further specifics depending on the Windows version or the like.
I see you point with having a central localization file. The problem is, however, that you usually want to assign a meaning to special characters (e.g., which kind of separator they are), which conflicts with having them in a central localization file, as then everyone has to depend on a project which has to contain information about other projects it does not depend on (and then reduces comprehensibility for those who have to modify that file). But, of course, either way has its drawback(s).

I have moved the character to a dedicated static variable in ee05475. Without that, my argument of assigning a meaning to special characters would not even be valid 😉